### PR TITLE
box/lua: disallow changing box options directly

### LINKED ISSUE #2867: Raise on raw modifications of box.cfg values
```sh
tarantool
> box.cfg{}
> box.cfg.read_only
false
> box.cfg.read_only = true
> box.cfg.read_only
true
```

I'd expect 3rd command should raise and cfg table at all would be readonly

Updated (25.01.2023): RFC - https://www.notion.so/Raw-modifications-of-box-cfg-values-9f440d321f984389a22bcb3f81366ea7

### PR DESCRIPTION
**Pull request by ligurio**
Tarantool has a special table 'box.cfg' that includes configuration
parameters. User could view that table, but it's direct modification has
no effect - after assigning a new value it is actually "updated" but
actual value remains the same. Such behaviour is a counterintuitive for
our users and provides a bad experience.

Proposed patch change this behaviour: new value assigned to parameter
via direct access to table box.cfg raise an error.

Before the patch:
```
tarantool> box.cfg{}
<snipped>
tarantool> box.cfg.read_only=true
---
...

tarantool>
```

After the patch:

```
tarantool> box.cfg{}
<snipped>
tarantool> box.cfg.read_only=true
---
- error: 'builtin/box/load_cfg.lua:973: Use function box.cfg for modifications'
...

tarantool>
```

NO_DOC=bugfix

Closes #2867

### CODE REVIEW

**Comment by Gumix** [comment]
Hi, here is something wrong with the validation of the options:
```lua
tarantool> box.cfg{}
-- skipped --

tarantool> box.cfg.log_format = 'xxx'
---
- error: 'Incorrect value for option ''log_format'': expected ''plain'' or ''json'''
...

tarantool> box.cfg.log_format = 'xxx'
---
...

tarantool> box.cfg.log_format = 'yyy'
---
...

tarantool> box.cfg.log_format
---
- yyy
...

tarantool> box.cfg.log_level = 6
---
- error: 'Incorrect value for option ''log_format'': expected ''plain'' or ''json'''
...
-- ??? why log_format?
```

Here is another issue, `box.cfg.log_level` is updated correctly, while `box.cfg` still prints the old value:
```lua
tarantool> box.cfg{}
-- skipped --

tarantool> box.cfg.log_level
---
- 5
...

tarantool> box.cfg.log_level = 6
2023-01-25 10:53:43.764 [339315] main/103/interactive/box.load_cfg I> set 'log_level' configuration option to 6
---
...

tarantool> box.cfg.log_level
---
- 6
...

tarantool> box.cfg
---
- replication_sync_timeout: 0
  -- skipped --
  log_level: 5
  -- skipped --
```

**Comment by igormunkin** [comment]
@Gumix, nice catch, thanks for noticing! By the way, I've added you to reviewers list, since we touch box-related interface. I believe you're the best candidate, 'cause you're the only one from Core team, who has already touched the patch :)

File: src/box/lua/load_cfg.lua
Comment (Gumix) [inline src/box/lua/load_cfg.lua:994]: Commented-out code.

**Review by Gumix — APPROVED** [review]
OK for me, but I want to invite @nshy

File: test/box-luatest/gh_2867_raise_on_raw_box_cfg_modifications_test.lua
Comment (nshy) [inline test/box-luatest/gh_2867_raise_on_raw_box_cfg_modifications_test.lua:38]: Do we need to drop this this test as it is unconditionally skipped?

File: src/box/lua/load_cfg.lua
Comment (nshy) [inline src/box/lua/load_cfg.lua:1000]: Great we finally going to fix this issue! I have a couple of concerns though.

1. We have config options of type table. Examples are `wal_ext` or `log_level_modules`. Can we make table proxies recursively so that changing nested tables directly is also prohibited?

2. We probably better make same proxy for log.cfg in this patch or this patch series too. log.cfg have same issue.

**Review by nshy — CHANGES_REQUESTED** [review]

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:1000]: Looks like we can just apply same approach with empty proxy table recursively on box.cfg. This way we make box.cfg readonly on all levels. Then we can put this functionality info function and apply to log.cfg too.

Comment (nshy) [reply src/box/lua/load_cfg.lua:1000]: Discussed f2f and decided to create a new task for the above issues.

**Review by nshy — APPROVED** [review]

File: test/box-luatest/gh_2867_raise_on_raw_box_cfg_modifications_test.lua
Comment (igormunkin) [inline test/box-luatest/gh_2867_raise_on_raw_box_cfg_modifications_test.lua:28]: Side note: nice "unknown" option ;)

File: src/lua/init.lua
Comment (igormunkin) [inline src/lua/init.lua:40]: Well, this part is not quite right.

1. It's unsafe, since `getmetatable` yields the value of `__metatable` field if one is set (that can be e.g. string[^string]). You need either to check the type of `mt` variable or to use `debug.getmetatable` to obtain `object`'s metatable. I'd rather use the latter approach.
[^string]: Fortunately, all Lua strings share the metatable with overloaded index, so `mt.__name` in case of strings just yields `nil`, but anyway in general case this is not safe.
2. There is no need in fetching `object`'s metatable one more time in the line above (it's already stored in `mt`, isn't it?) so just use `mt.__index` for this.

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [inline src/box/lua/load_cfg.lua:972]: Typo: s/existed/existing/.

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [inline src/box/lua/load_cfg.lua:991]: Why do we need to differ these two cases? I believe `box.cfg` modification is forbidden, whether the key exists in the table or not.

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [inline src/box/lua/load_cfg.lua:804]: Why all these three changes above are required? AFAIU, `oldcfg` parameter is `raw_cfg` as a result of your patch and this table has no metatable, hence the change has no effect. However, the code in `load_cfg.lua` is not quite clear, so I might be wrong.

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [inline src/box/lua/load_cfg.lua:1123]: The same as for the three similar hunks above.

**Review by igormunkin — CHANGES_REQUESTED** [review]
@ligurio, thanks for the patch! Please consider my comments below. 

Furthermore, there are two confusing `NOWRAP` entries in the commit message: if these are just typos, please fix it either.

I also guess that adding a docbot request will be nice, since such behaviour change should be added to the docs.

**Comment by Totktonada** [comment]
I glanced over the patch out of curiosity and would give a few suggestions about the code layout.

* I would move our own `pairs()` into its own module src/lua/pairs.lua (just like we do for `print()`). src/lua/init.lua is bloated with many things unrelated to each other and I think that it should be decomposed in a future anyway.
* I would eliminate the `__name` field and would the tables itself as keys in a weak table... well, it is easier to explain with code:

```lua
-- pairs.lua
local M = {}

M.raw_pairs = pairs

-- table -> iterator generator
M.tracked = setmetatable({}, {__mode = 'k'})

_G.pairs = function(t)
    local iter_gen = M.tracked[t] or M.raw_pairs
    return iter_gen(t)
end

function M.add(t, iter_gen)
    M.tracked[t] = iter_gen
end

function M.remove(t)
    M.tracked[t] = nil
end

return M
```

Usage:

```lua
local internal_pairs = require('internal.pairs')
local real_table = {}
local proxy_table = {_iterate_over = real_table}
internal_pairs.add(proxy_table, function(self)
    return pairs(self._iterate_over)
end)

-- or

local internal_pairs = require('internal.pairs')
local real_table = {}
local proxy_table = {}
internal_pairs.add(proxy_table, function(_self)
    return pairs(real_table)
end)
```

Feel free to ignore, I just pass here over.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:804]: `rawset` is not required anymore because internally we start to use `raw_cfg`. So I would drop `rawset` too.

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [reply src/box/lua/load_cfg.lua:1045]: Yeah, nice change, thanks.

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [reply src/box/lua/load_cfg.lua:875]: Great, thanks!

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [reply src/box/lua/load_cfg.lua:1177]: Yeah, that one is definitely required. Anyway, I see the second patch totally removes `rawset` from `load_cfg.lua`, and that's neat. Thanks!

File: src/box/lua/load_cfg.lua
Comment (igormunkin) [inline src/box/lua/load_cfg.lua:1039]: Minor: The only thing that bothers me here is the condition.

Is it possible to pass other "compex" types (such as cdata/udata/function) as a `box.cfg` parameter value? If not, then everything is fine; if there are doubts we can just leave it as is and fix if/when one is found: yeah, it's bad when there is garbage in the error message, but not critical.

Feel free to ignore.

**Review by igormunkin — APPROVED** [review]
@ligurio, thanks for all changes made! LGTM now (but I'll wait for your answer for my question below prior to merging the PR).

**Pull request merged**
